library/pkcs7.c

```diff
@@ -169,6 +169,91 @@ static int pkcs7_get_digest_algorithm_set(unsigned char **p,
     return 0;
 }
 
+/**
+ * SignedAttributes ::= SET SIZE (1..MAX) OF Attribute
+ *
+ * Attribute ::= SEQUENCE {
+ *      attrType OBJECT IDENTIFIER,
+ *      attrValues SET OF AttributeValue }
+ **/
+static int pkcs7_get_signed_attrs(unsigned char **p,
+                                  unsigned char *end,
+                                  mbedtls_x509_buf *attrs_raw,
+                                  mbedtls_asn1_named_data *attrs)
+{
+    mbedtls_asn1_named_data *cur = attrs;
+    unsigned char *seq_end;
+    size_t len;
+    int ret = 0;
+
+    attrs_raw->p = *p;
+    attrs_raw->len = 0;
+
+    ret = mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                               | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+    if (ret != 0) {
+        *p = attrs_raw->p;
+        attrs_raw->p = NULL;
+        return 0;
+    }
+    end = *p + len;
+
+    while (*p < end) {
+        ret = mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                                   | MBEDTLS_ASN1_SEQUENCE);
+        if (ret != 0) {
+            goto err;
+        }
+        seq_end = *p + len;
+
+        ret = mbedtls_asn1_get_tag(p, seq_end, &len, MBEDTLS_ASN1_OID);
+        if (ret != 0) {
+            goto err;
+        }
+        cur->oid.p = *p;
+        cur->oid.len = len;
+        *p += len;
+
+        ret = mbedtls_asn1_get_tag(p, seq_end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                                   | MBEDTLS_ASN1_SET);
+        if (ret != 0) {
+            goto err;
+        }
+        cur->val.tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SET;
+        cur->val.p = *p;
+        cur->val.len = len;
+        *p += len;
+
+        if (*p != seq_end) {
+            ret = MBEDTLS_ERR_ASN1_LENGTH_MISMATCH;
+            goto err;
+        }
+        if (*p != end) {
+            cur->next = mbedtls_calloc(1, sizeof(*cur->next));
+            if (cur->next == NULL) {
+                ret = MBEDTLS_ERR_PKCS7_ALLOC_FAILED;
+                goto err;
+            }
+            cur = cur->next;
+        }
+    }
```

Comment:
X.690 requires that `SET OF` types be sorted (IIUC first by length, then by value if lengths are equal), but this code doesn’t check for that.  Mbed TLS should really have an `mbedtls_asn1_traverse_set_of()` utility function that checks that the values are sorted and that the callback used all bytes in each entry.  Additionally, the code should check that `MBEDTLS_OID_PKCS9_MESSAGE_DIGEST` is present exactly once.